### ISSUE TITLE
Regex compilation bug when prefix is shared by regexes

### ISSUE THREAD

**Issue by osa1** [osa1/lexgen#16]
```rust
use lexgen::lexer;

fn ignore_pos<A, E>(ret: Option<Result<(usize, A, usize), E>>) -> Option<Result<A, E>> {
    ret.map(|res| res.map(|(_, a, _)| a))
}

fn next<A, E>(
    iter: &mut dyn Iterator<Item = Result<(usize, A, usize), E>>,
) -> Option<Result<A, E>> {
    ignore_pos(iter.next())
}

fn return_match<'lexer, 'input>(lexer: LexerHandle<'lexer, 'input>) -> LexerAction<&'input str> {
    let match_ = lexer.match_();
    lexer.return_(match_)
}

lexer! {
    Lexer -> &'input str;

    "xyzxyz" => return_match,
    "xyz" => return_match,
    "xya" => return_match,
}

fn main() {
    let mut lexer = Lexer::new("xyzxya");
    assert_eq!(next(&mut lexer), Some(Ok("xyz")));  // fails
    assert_eq!(next(&mut lexer), Some(Ok("xya")));
    assert_eq!(next(&mut lexer), None);
}
```

Interestingly, we have very similar examples that work fine. For example, Lua lexer has regexes for "elseif" and "else".

**Comment by osa1** [comment]
It seems like ocamllex implements this with backtracking. For the program above, if I add some prints in generated code and try it with `"aaaaaaaa"` I see this output:

```
state 4 - a seen, curr=3, last=1
state 4 - a seen, curr=4, last=1
state 4 - a seen, curr=5, last=1
state 4 - a seen, curr=6, last=1
state 4 - a seen, curr=7, last=1
state 4 - a seen, curr=8, last=1
EOF reached in state 4, curr=1, last=1
state 4 - a seen, curr=4, last=2
state 4 - a seen, curr=5, last=2
state 4 - a seen, curr=6, last=2
state 4 - a seen, curr=7, last=2
state 4 - a seen, curr=8, last=2
EOF reached in state 4, curr=2, last=2
state 4 - a seen, curr=5, last=3
state 4 - a seen, curr=6, last=3
state 4 - a seen, curr=7, last=3
state 4 - a seen, curr=8, last=3
EOF reached in state 4, curr=3, last=3
state 4 - a seen, curr=6, last=4
state 4 - a seen, curr=7, last=4
state 4 - a seen, curr=8, last=4
EOF reached in state 4, curr=4, last=4
state 4 - a seen, curr=7, last=5
state 4 - a seen, curr=8, last=5
EOF reached in state 4, curr=5, last=5
state 4 - a seen, curr=8, last=6
EOF reached in state 4, curr=6, last=6
EOF reached in state 4, curr=7, last=7
```

So it handles one match at a time, and scans the prefix again every time we see EOF instead of 'b'.

Output when I run it on `"aaaaaaaa"`:

```
state 4 - a seen, curr=3, last=1
state 4 - a seen, curr=4, last=1
state 4 - a seen, curr=5, last=1
state 4 - a seen, curr=6, last=1
state 4 - a seen, curr=7, last=1
state 4 - b seen, curr=8, last=1
```

<details>

<summary>ocamllex-generated code with prints</summary>

```ocaml
# 1 "lexer.mll"
  
# 4 "lexer.ml"

let rec __ocaml_lex_refill_buf lexbuf _buf _len _curr _last =
  if lexbuf.Lexing.lex_eof_reached then
    256, _buf, _len, _curr, _last
  else begin
    lexbuf.Lexing.lex_curr_pos <- _curr;
    lexbuf.Lexing.lex_last_pos <- _last;
    lexbuf.Lexing.refill_buff lexbuf;
    let _curr = lexbuf.Lexing.lex_curr_pos in
    let _last = lexbuf.Lexing.lex_last_pos in
    let _len = lexbuf.Lexing.lex_buffer_len in
    let _buf = lexbuf.Lexing.lex_buffer in
    if _curr < _len then
      Char.code (Bytes.unsafe_get _buf _curr), _buf, _len, (_curr + 1), _last
    else
      __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
  end

let rec __ocaml_lex_state4 lexbuf _last_action _buf _len _curr _last =

  let next_char, _buf, _len, _curr, _last =
    if _curr >= _len then
      __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
    else
      Char.code (Bytes.unsafe_get _buf _curr),
      _buf, _len, (_curr + 1), _last
  in
  begin match next_char with
    (* |'a' *)
    | 97 ->
      Printf.printf "state 4 - a seen, curr=%d, last=%d\n" _curr _last;
      __ocaml_lex_state4 lexbuf _last_action _buf _len _curr _last
    (* |'b' *)
    | 98 ->
      Printf.printf "state 4 - b seen, curr=%d, last=%d\n" _curr _last;
      (* *)
      lexbuf.Lexing.lex_curr_pos <- _curr;
      lexbuf.Lexing.lex_last_pos <- _last;
      0
    | _ ->
      let _curr = _last in
      Printf.printf "EOF reached in state 4, curr=%d, last=%d\n" _curr _last;
      lexbuf.Lexing.lex_curr_pos <- _curr;
      lexbuf.Lexing.lex_last_pos <- _last;
      _last_action
  end


let rec rule1 lexbuf =
  let __ocaml_lex_result =
    let _curr = lexbuf.Lexing.lex_curr_pos in
    let _last = _curr in
    let _len = lexbuf.Lexing.lex_buffer_len in
    let _buf = lexbuf.Lexing.lex_buffer in
    let _last_action = -1 in
    lexbuf.Lexing.lex_start_pos <- _curr;
    let next_char, _buf, _len, _curr, _last =
      if _curr >= _len then
        __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
      else
        Char.code (Bytes.unsafe_get _buf _curr),
        _buf, _len, (_curr + 1), _last
    in
    begin match next_char with
      (* |'a' *)
      | 97 ->
        (* *)
        let _last = _curr in
        let _last_action = 1 in
        let next_char, _buf, _len, _curr, _last =
          if _curr >= _len then
            __ocaml_lex_refill_buf lexbuf _buf _len _curr _last
          else
            Char.code (Bytes.unsafe_get _buf _curr),
            _buf, _len, (_curr + 1), _last
        in
        begin match next_char with
          (* |'a' *)
          | 97 ->
            __ocaml_lex_state4 lexbuf 1 (* = last_action *) _buf _len _curr _last
          (* |'b' *)
          | 98 ->
            (* *)
            lexbuf.Lexing.lex_curr_pos <- _curr;
            lexbuf.Lexing.lex_last_pos <- _last;
            0
          | _ ->
            let _curr = _last in
            lexbuf.Lexing.lex_curr_pos <- _curr;
            lexbuf.Lexing.lex_last_pos <- _last;
            1 (* = last_action *)
        end
      (* |eof *)
      | 256 ->
        (* *)
        lexbuf.Lexing.lex_curr_pos <- _curr;
        lexbuf.Lexing.lex_last_pos <- _last;
        2
      | _ ->
        let _curr = _last in
        lexbuf.Lexing.lex_curr_pos <- _curr;
        lexbuf.Lexing.lex_last_pos <- _last;
        _last_action
    end
  in
  begin
    let _curr_p = lexbuf.Lexing.lex_curr_p in
    if _curr_p != Lexing.dummy_pos then begin
      lexbuf.Lexing.lex_start_p <- _curr_p;
      lexbuf.Lexing.lex_curr_p <-
        {_curr_p with Lexing.pos_cnum =
         lexbuf.Lexing.lex_abs_pos+lexbuf.Lexing.lex_curr_pos}
    end
  end;
  match __ocaml_lex_result with
  | 0 ->
# 4 "lexer.mll"
             ( rule1 lexbuf )
# 119 "lexer.ml"

  | 1 ->
# 5 "lexer.mll"
             ( rule1 lexbuf )
# 124 "lexer.ml"

  | 2 ->
# 6 "lexer.mll"
             ( () )
# 129 "lexer.ml"

  | _ -> raise (Failure "lexing: empty token")


;;

# 8 "lexer.mll"
 
  (*
  let time f x =
    let start = Unix.gettimeofday () in
    let res = f x in
    let stop = Unix.gettimeofday () in
    let () = Printf.printf "%fs\n%!" (stop -. start) in
    res
  in

  let str0 = String.make 10000 'a' in
  let str1 = String.make 20000 'a' in
  let str2 = String.make 40000 'a' in
  let str3 = String.make 80000 'a' in
  let str4 = String.make 160000 'a' in
  let str5 = String.make 320000 'a' in

  time (fun () -> rule1 (Lexing.from_string str0)) ();
  time (fun () -> rule1 (Lexing.from_string str1)) ();
  time (fun () -> rule1 (Lexing.from_string str2)) ();
  time (fun () -> rule1 (Lexing.from_string str3)) ();
  time (fun () -> rule1 (Lexing.from_string str4)) ();
  time (fun () -> rule1 (Lexing.from_string str5)) ()
  *)

  rule1 (Lexing.from_string "aaaaaaab")

# 164 "lexer.ml"
```

</details>

**Comment by osa1** [comment]
I started fixing this in branch `backtracking`. So far I implemented backtracking NFA simulation. Added the `a+b | a` example above as a test and it works.

I'm not sure if in the generated code we want to do backtracking, but it's good enough for NFA and DFA simulations. I'm guessing if it's good enough for ocamllex it could be good enough for us too, so maybe I will also implement backtracking in the generated code.
